### PR TITLE
Expose coordinate bounds methods

### PR DESCRIPTION
**Pull request by evil159**
This PR exposes some utility methods/properties for `CoordinateBounds` to public.

Additionally to the methods outlined in https://github.com/mapbox/mapbox-maps-ios/issues/207, `east`, `west`, `north`, `south`, `isEmpty` were exposed together with equality and hashability support.

Fixes: https://github.com/mapbox/mapbox-maps-ios/issues/207

## Pull request checklist:
 - [x] Write tests for all new functionality. If tests were not written, please explain why.
 - [x] Add documentation comments for any added or updated public APIs.
 - [x] Describe the changes in this PR, especially public API changes.
 - [x] Add a changelog entry to to bottom of the relevant section (typically the `## main` heading near the top).
 - [x] Review and agree to the Contributor License Agreement ([CLA](https://github.com/mapbox/mapbox-maps-ios/blob/main/CONTRIBUTING.md#contributor-license-agreement)).

### CODE REVIEW

File: Sources/MapboxMaps/Foundation/Extensions/Core/CoordinateBounds.swift [hunk 1]
```diff
@@ -5,27 +5,106 @@ import MapboxCoreMaps
...
+
+    /// Returns whether the bounds are empty or not.
+    var isEmpty: Bool { __isEmpty() }
+
+    /// The absolute distance, in degrees, between the north and south boundaries of these bounds.
+    ///
+    /// One degree of latitude is approximately 111 kilometers (69 miles).
+    var latitudeSpan: CLLocationDegrees {
+        __latitudeSpan()
+    }
```
Comment (macdrevx) [inline Sources/MapboxMaps/Foundation/Extensions/Core/CoordinateBounds.swift:34]: inconsistent formatting. please be consistent about whether to include `return` and whether to wrap or write all on one line

File: Sources/MapboxMaps/Foundation/Extensions/Core/CoordinateBounds.swift [hunk 2]
```diff
@@ -5,27 +5,106 @@ import MapboxCoreMaps
...
-        for coordinate in coordinates {
-            if latitudeRange.contains(coordinate.latitude) || longitudeRange.contains(coordinate.longitude) {
-                return true
-            }
+    /// Extends these bounds to include `point`.
+    /// - Parameter coordinate: The coordinate to be included.
+    /// - Returns: The extended bounds.
+    func include(_ coordinate: CLLocationCoordinate2D) -> CoordinateBounds {
+        extend(forPoint: coordinate)
+    }
```
Comment (macdrevx) [inline Sources/MapboxMaps/Foundation/Extensions/Core/CoordinateBounds.swift:67]: Can you confirm whether these functions return new instances or mutate self?

I'd also like to workshop on these names with you a bit to think through consistency with Android and adherence to the Swift API Design Guidelines.

File: Sources/MapboxMaps/Foundation/Extensions/Core/CoordinateBounds.swift [hunk 3]
```diff
@@ -5,27 +5,106 @@ import MapboxCoreMaps
...
+
+    override func isEqual(_ object: Any?) -> Bool {
+        guard let other = object as? CoordinateBounds else {
+            return false
         }
-        return false
+
+        return self == other
+    }
```
Comment (macdrevx) [inline Sources/MapboxMaps/Foundation/Extensions/Core/CoordinateBounds.swift:102]: I'm wary of overriding methods in extensions. Is it safe?

Comment (macdrevx) [reply Sources/MapboxMaps/Foundation/Extensions/Core/CoordinateBounds.swift:102]: Seems it's allowed only for extensions of Objective-C types. What's the correct mental model for how it works? What happens if the same type has its own implementation?

File: Sources/MapboxMaps/Foundation/Extensions/Core/CoordinateBounds.swift (hunk 2 above)
Comment (evil159) [reply Sources/MapboxMaps/Foundation/Extensions/Core/CoordinateBounds.swift:67]: They return new instances.

Yes, the inconsistency between `union` and `include` is puzzling, given that they perform similar things.

File: Sources/MapboxMaps/Foundation/Extensions/Core/CoordinateBounds.swift [hunk 4]
```diff
@@ -5,27 +5,106 @@ import MapboxCoreMaps
...
+    /// - Parameter coordinate: The coordinate to be included.
+    /// - Returns: The extended bounds.
+    func include(_ coordinate: CLLocationCoordinate2D) -> CoordinateBounds {
+        extend(forPoint: coordinate)
+    }
+
+    /// Returns a new `CoordinateBounds` that is the intersection of this with another `CoordinateBounds`.
+    /// - Parameter coordinateBounds: The `CoordinateBounds` to intersect with.
+    /// - Returns: The bounds representing the intersection of the two bounds or null if there is no intersection.
+    func intersect(_ coordinateBounds: CoordinateBounds) -> CoordinateBounds? {
```
[the remark this answers is not in the document]
Comment (evil159) [reply Sources/MapboxMaps/Foundation/Extensions/Core/CoordinateBounds.swift:72]: It's here https://github.com/mapbox/mapbox-gl-native-android/blob/d25d1335f97cfcdbb45deab50d384cc4a5ecf8ad/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/geometry/LatLngBounds.java#L454

[the remark this answers is not in the document]
Comment (macdrevx) [reply Sources/MapboxMaps/Foundation/Extensions/Core/CoordinateBounds.swift:62]: I believe that's a link to android v9

File: Sources/MapboxMaps/Foundation/Extensions/Core/CoordinateBounds.swift (hunk 2 above)
Comment (evil159) [reply Sources/MapboxMaps/Foundation/Extensions/Core/CoordinateBounds.swift]: Removed as a result of our workshop.

File: Sources/MapboxMaps/Foundation/Extensions/Core/CoordinateBounds.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (evil159) [reply Sources/MapboxMaps/Foundation/Extensions/Core/CoordinateBounds.swift]: 🤦

File: Sources/MapboxMaps/Foundation/Extensions/Core/CoordinateBounds.swift (hunk 3 above)
Comment (evil159) [reply Sources/MapboxMaps/Foundation/Extensions/Core/CoordinateBounds.swift]: Not that safe it seems:

> If the name of a method declared in a category is the same as a method in the original class, or a method in another category on the same class (or even a superclass), the behavior is undefined as to which method implementation is used at runtime. This is less likely to be an issue if you’re using categories with your own classes, but can cause problems when using categories to add methods to standard Cocoa or Cocoa Touch classes.

From [here](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ProgrammingWithObjectiveC/CustomizingExistingClasses/CustomizingExistingClasses.html), removed them.

**Review by macdrevx — APPROVED** [review]

**Pull request merged**